File: advent/day01.py
# /usr/bin/env python3

from advent.base import BaseSolver, Solution


class Solver(BaseSolver):
    NUMS = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]

    def solve(self) -> Solution:
        res1 = 0
        res2 = 0
        for line in self.data.splitlines():
            nums1 = []
            nums2 = []
            for i, c in enumerate(line):
                if c.isdigit():
                    nums1.append(c)
                    nums2.append(c)

                for n in self.NUMS:
                    if line[i:].startswith(n):
                        nums2.append(str(self.NUMS.index(n) + 1))
            res1 += int(nums1[0] + nums1[-1])
            res2 += int(nums2[0] + nums2[-1])
        return Solution(res1, res2)


Solver.run()

File: nim/day01.nim
import strutils
import tables

type Solution = tuple[part1: int, part2: int]

let NUMS = {
  "one": '1', 
  "two": '2',
  "three": '3',
  "four": '4',
  "five": '5',
  "six": '6',
  "seven": '7',
  "eight": '8',
  "nine": '9',
}.toTable

proc solve(): Solution =
  var
    res1 = 0
    res2 = 0

  for line in stdin.lines:
    var
      nums1: seq[char] = @[]
      nums2: seq[char] = @[]

    for i, c in line:
      if c.isDigit():
        nums1.add(c)
        nums2.add(c)

      for s, c in NUMS:
        if line[i ..< line.len].startsWith(s):
          nums2.add(c)
    res1 += (nums1[0] & nums1[^1]).parseInt
    res2 += (nums2[0] & nums2[^1]).parseInt

  (res1, res2)

echo solve()
